### DISCUSSION TITLE
Announcing Unirest 4 & Upgrade Guide

### DISCUSSION THREAD

**Discussion by ryber** [Kong/unirest-java#429] [Show and tell]
## Whats New?
* Full support for HTTP/2
* Websocket Support
* Modular Plugins for JSON and other extensions. Unirest now comes with ZERO dependencies by default
* Dependency management BOM

## Upgrading to Unirest 4.0

***Warning: Unirest 4 is still in release candidate phase and could still have breaking changes introduced***

 🚨 **Unirest 4 drops the Apache Http Client dependency in favor of the pure Java client** 🚨. This means that unirest 4 has a dependency on Java-11. Unirest 3 will continue to be supported for bugs and minor features. If you still haven't upgraded from Java 8 now is the time!  

Due to engine changes there are some differences in behavior. Efforts have been made to limit these changes. The others are documented here.

### How to upgrade
* The core Unirest package has been moved from ```kong.unirest``` to ```kong.unirest.core```, this prevents classloader issues where Unirest 4 may conflict with previous versions which may also be on the classpath.
* Unirest no longer comes with GSON as a default JSON parser. In order to enable JSON support you must include either the GSON or Jackson JSON packages with:
```xml
<dependency>
  <groupId>com.konghq</groupId>
  <artifactId>unirest-object-mappers-gson</artifactId>
  <version>4.0.0-RC4</version>
</dependency>
<!-- or -->
<dependency>
  <groupId>com.konghq</groupId>
  <artifactId>unirest-object-mappers-jackson</artifactId>
  <version>4.0.0-RC4</version>
</dependency>
```
* In order to assist with the various modules Unirest now includes a dependency management BOM. Include the BOM in your dependency management section and then just declare the modules you want without the version.

```xml
<dependencyManagement>
  <dependencies>
      <dependency>
          <groupId>com.konghq</groupId>
          <artifactId>unirest-java-bom</artifactId>
          <version>4.0.0-RC4</version>
          <type>pom</type>
          <scope>import</scope>
      </dependency>
  </dependencies>
</dependencyManagement>

<dependencies>
  <dependency>
    <groupId>com.konghq</groupId>
    <artifactId>unirest-java-core</artifactId>
  </dependency>
  <dependency>
    <groupId>com.konghq</groupId>
    <artifactId>unirest-object-mappers-gson</artifactId>
  </dependency>
</dependencies>
```

### Differences
* null header values are now represented by an empty string rather than null.
* response headers may return a different order that with Apache.
* You may not normally override the `host` header anymore. Starting with Java-12 you may only due this by setting a system property of jdk.httpclient.allowRestrictedHeaders=host. see https://bugs.openjdk.java.net/browse/JDK-8213696
* Cookie management follows more modern standards and may differ from apache with regard to non-standard parsing. 
   * related: ```config.cookieSpec(String)``` has been removed as it was Apache specific.
* Per-Request proxies are no longer supported.  
* Custom HostNameVerifier is no longer supported.
* Socket timeout is no longer set independent of connection timeout and has been removed
* There are no longer any monitoring threads to shut down, as such, all close methods and the registering of shutdown hooks have been removed.
* Using system props for proxy settings is false by default (was true in previous versions)

**Comment by Rakambda** [comment]
Dependency seems to be
```xml
<dependency>
  <groupId>com.konghq</groupId>
  <artifactId>unirest-java-core</artifactId>
</dependency>
  ```
  instead of
  ```xml
<dependency>
    <groupId>com.konghq</groupId>
    <artifactId>unirest-java</artifactId>
</dependency>
  ```

**Reply by ryber** [reply]
Corrected

**Reply by ryber** [reply]
theres also a bug in the dependency management bom I'm working on fixing. Ooops, little premature!

**Comment by GeorgeSalu** [comment]
is there a roadmap for version 4? Do you have a forecast for release of version 4?

**Reply by ryber** [reply]
I'm planning on releasing the first GA version around Christmas. There are issues for roadmap stuff in the issues section

**Comment by mikymigs** [comment]
Hello,

We would like to migrate to Unirest 4 but we're facing a major challenge with the change from Apache Http Client to Java Http Client:

- We use Zipkin / Brave for tracing and there seems to be no implementation for Java Http Client. We're thinking of maybe moving to a different tracing tool, would you have any suggestion best suited with Unirest support ?
- We have request/response interceptors that we use for debug logs and it seems that the Java Http Client does not propose an easy way to declare interceptors (looks like you basically have to code the logic from scratch). Is there any request/response interceptor logic implemented in Unirest that we may be unaware of or how do you handle such a need with Java Http Client?

Thanks a lot,

**Reply by ryber** [reply]
1. Unirest doesn't have any code specific to any tracing tools. You can set global default headers with a supplier which should work with most of them. I have no personal opinion on any of them.
2. Yes there is an interceptor. You can also create a [single one](https://github.com/Kong/unirest-java/blob/main/unirest-bdd-tests/src/test/java/BehaviorTests/InterceptorTest.java) or create a [compound one with multiple steps](https://github.com/Kong/unirest-java/blob/main/unirest/src/test/java/kong/unirest/core/CompoundInterceptorTest.java)

**Reply by mikymigs** [reply]
Thank you for your answer, it helps us taking the right direction. 
We'll have a look at the suggested options.
